### ISSUE TITLE
Unable to build firefox or chrome due to SSL interpreting warnings as errors.

### ISSUE THREAD

**Issue by SuspiciousDuck** [lwthiker/curl-impersonate#201]
I am using Arch and was attempting to build libcurl-impersonate from source. However, the both the [firefox](https://aur.archlinux.org/packages/curl-impersonate-firefox) and [chrome](https://aur.archlinux.org/packages/curl-impersonate-chrome) AUR packages keep failing to build, even cloning the repo and building it manually doesn't build right. Oddly enough, its the SSL packages that keep failing to build. The error message I get when building firefox is:
```
FAILED: obj/lib/ssl/ssl.ssl3exthandle.o
cc -MMD -MF obj/lib/ssl/ssl.ssl3exthandle.o.d -DNSS_FIPS_DISABLED -DNSS_NO_INIT_SUPPORT -DNSS_X86_OR_X64 -DNSS_X64 -DNSS_USE_64 -DNSS_ALLOW_SSLKEYLOGFILE -DUSE_UTIL_DIRECTLY -DNO_NSPR_10_SUPPORT -DSSL_DISABLE_DEPRECATED_CIPHER_SUITE_NAMES -DLINUX2_1 -DLINUX -Dlinux -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_POSIX_SOURCE -DSDB_MEASURE_USE_TEMP_DIR -DHAVE_STRERROR -DXP_UNIX -D_REENTRANT -DNSS_DISABLE_DBM -DNSS_DISABLE_LIBPKIX -DNDEBUG -I/home/xxxx/.cache/yay/curl-impersonate-firefox/src/curl-impersonate-0.5.4/build/nss-3.87/dist/Release/include/nspr -I/home/xxxx/.cache/yay/curl-impersonate-firefox/src/curl-impersonate-0.5.4/build/nss-3.87/dist/private/nss -I/home/xxxx/.cache/yay/curl-impersonate-firefox/src/curl-impersonate-0.5.4/build/nss-3.87/dist/public/nss -fPIC -pipe -ffunction-sections -fdata-sections -m64 -Werror -Wall -Wshadow -O2 -std=c99  -c ../../lib/ssl/ssl3exthandle.c -o obj/lib/ssl/ssl.ssl3exthandle.o
../../lib/ssl/ssl3exthandle.c:205:1: error: conflicting types for ‘ssl3_ClientSendSessionTicketXtn’ due to enum/integer mismatch; have ‘PRInt32(const sslSocket *, TLSExtensionData *, sslBuffer *, PRBool *)’ {aka ‘int(const struct sslSocketStr *, struct TLSExtensionDataStr *, struct sslBufferStr *, int *)’} [-Werror=enum-int-mismatch]
  205 | ssl3_ClientSendSessionTicketXtn(const sslSocket *ss, TLSExtensionData *xtnData,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../lib/ssl/ssl3exthandle.c:17:
../../lib/ssl/ssl3exthandle.h:116:11: note: previous declaration of ‘ssl3_ClientSendSessionTicketXtn’ with type ‘SECStatus(const sslSocket *, TLSExtensionData *, sslBuffer *, PRBool *)’ {aka ‘enum _SECStatus(const struct sslSocketStr *, struct TLSExtensionDataStr *, struct sslBufferStr *, int *)’}
  116 | SECStatus ssl3_ClientSendSessionTicketXtn(const sslSocket *ss,
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```
and the error message when building chrome is:
```
FAILED: ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o
/usr/bin/g++ -DBORINGSSL_IMPLEMENTATION -I/home/xxxx/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.5.4/build/boringssl/third_party/googletest/include -I/home/xxxx/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.5.4/build/boringssl/ssl/test/../../include -Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wvla -Wshadow -ggdb -Wall -fvisibility=hidden -fno-common -Wno-free-nonheap-object -Wimplicit-fallthrough -Wmissing-declarations -std=c++11 -fno-exceptions -fno-rtti -O3 -DNDEBUG -fPIE -MD -MT ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o -MF ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o.d -o ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o -c /home/xxxx/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.5.4/build/boringssl/ssl/test/settings_writer.cc
/home/xxxx/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.5.4/build/boringssl/ssl/test/settings_writer.cc: In member function ‘bool SettingsWriter::Commit()’:
/home/xxxx/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.5.4/build/boringssl/ssl/test/settings_writer.cc:78:61: error: ignoring attributes on template argument ‘int (*)(FILE*)’ [-Werror=ignored-attributes]
   78 |   using ScopedFILE = std::unique_ptr<FILE, decltype(&fclose)>;
      |                                                             ^
cc1plus: all warnings being treated as errors
```
Is there a way to fix this? I can't seem to get a working command to replace the flags with sed, but I hacked together a command to find matches using [The Silver Searcher](https://github.com/ggreer/the_silver_searcher).
```bash
$ ag ' -Wall| -Werror| -Wno-error' -l ~/curl-impersonate/
~/curl-impersonate/tests/Dockerfile
~/curl-impersonate/build/nss-3.87/nss/doc/rst/legacy/nss_releases/nss_3.55_release_notes/index.rst
~/curl-impersonate/build/nss-3.87/nss/doc/rst/legacy/nss_3.12_release_notes.html/index.rst
~/curl-impersonate/build/nss-3.87/nss/coreconf/NetBSD.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/SunOS4.1.3_U1.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/ReliantUNIX.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/NCR3.0.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/OpenBSD.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/BSD_OS.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/OS2.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/QNX.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/FreeBSD.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/Werror.mk
~/curl-impersonate/build/nss-3.87/nss/coreconf/SunOS5.mk
~/curl-impersonate/build/nss-3.87/nss/gtests/common/gtest.mk
~/curl-impersonate/build/nss-3.87/nss/gtests/google_test/gtest/cmake/internal_utils.cmake
~/curl-impersonate/build/nss-3.87/nss/lib/freebl/mpi/README
~/curl-impersonate/build/nss-3.87/nspr/configure.in
~/curl-impersonate/build/nss-3.87/nspr/configure
~/curl-impersonate/build/nss-3.87/nspr/pr/src/misc/dtoa.c
~/curl-impersonate/build/boringssl/CMakeLists.txt
~/curl-impersonate/build/boringssl/third_party/googletest/cmake/internal_utils.cmake
~/curl-impersonate/build/brotli-1.0.9/c/common/dictionary.c
```

**Comment by SuspiciousDuck** [comment]
Update: I was able to get Chrome compiling and working correctly after removing all appearances of -Wall and -Werror from /curl-impersonate/build/boringssl/CMakeLists.txt, but I haven't tried doing the same with firefox because I don't want to do all of that manually.
```bash
$ curl_chrome116 --version
curl 8.1.1 (x86_64-pc-linux-gnu) libcurl/8.1.1 BoringSSL zlib/1.3 brotli/1.0.9 zstd/1.5.5 libidn2/2.3.4 libpsl/0.21.2 (+libidn2/2.3.4) nghttp2/1.56.0
Release-Date: 2023-05-23
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM NTLM_WB PSL SSL threadsafe UnixSockets zstd
```

**Comment by SuspiciousDuck** [comment]
@lwthiker Possible fix could be cloning the repositories, replace all occurrences of -Wall and -Werror with a blank string, then building.

**Comment by lwthiker** [comment]
@SuspiciousDuck May you try out https://github.com/lwthiker/curl-impersonate/pull/217?
I upgraded the BoringSSL version to the latest stable one, and confirmed that it compiles on Arch without any error.
Would be great if you could give it a try as well.

**Comment by SuspiciousDuck** [comment]
@lwthiker Can I see how you tried doing it? Mine didn't turn out so well.
$ yay -S curl-impersonate-chrome --noconfirm
(...)
==> Starting build()...
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking for g++... g++
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... none needed
checking for strip... strip
checking for inflateEnd in -lz... yes
checking for cmake3... no
checking for cmake... cmake
checking for ninja... ninja
configure: creating ./config.status
config.status: creating Makefile
curl -L "https://github.com/google/brotli/archive/refs/tags/v1.0.9.tar.gz" \
	-o "brotli-1.0.9.tar.gz"
curl -L https://github.com/google/boringssl/archive/3a667d10e94186fd503966f5638e134fe9fb4080.zip \
	-o boringssl.zip
unzip -q -o boringssl.zip
mv boringssl-3a667d10e94186fd503966f5638e134fe9fb4080 boringssl
cd boringssl/
for p in /home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/../chrome/patches/boringssl-old-ciphers.patch; do patch -p1 < $p; done
touch .patched
curl -L https://github.com/nghttp2/nghttp2/releases/download/v1.56.0/nghttp2-1.56.0.tar.bz2 -o nghttp2-1.56.0.tar.bz2
curl -L "https://curl.se/download/curl-8.1.1.tar.xz" \
	-o "curl-8.1.1.tar.xz"
rm -Rf curl-8.1.1
tar -xf curl-8.1.1.tar.xz
cd curl-8.1.1
for p in /home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/../chrome/patches/curl-impersonate.patch; do patch -p1 < $p; done
\# Re-generate the configure script
autoreconf -fi
touch .patched-chrome
rm -f .patched-ff
unzip:  cannot find or open boringssl.zip, boringssl.zip.zip or boringssl.zip.ZIP.
make: *** [Makefile:261: boringssl/.patched] Error 9
make: *** Waiting for unfinished jobs....
tar: curl-8.1.1.tar.xz: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
make: *** [Makefile:346: curl-8.1.1/.patched-chrome] Error 2
      %%%   TTTooott taa all%l      T   o  % t%  a RlRe ec% ce  e iRi%v Receeveicedveed di % v%%  e XXfedX frf%ede rrd  Ave  drXA agefv  eeASrrvpedaer gea edgA  evS  ep SreTpaeiegdmee e  d    ST  pi  emTTeeiid  mme  e      T  Ti  im  meTTe ii  mm  ee     T   iT Cmi uem r reT e in  mt Ce
 u  Tr  irmC eeu nt r 
 r  Ce  un  rt  r
  e    n    t   
                                                                            D   l   o   a   d        D  Ul  po  la  od  U Dap lld oo  aa Ddd l  To  oa UtdTpal olo t aUa d Spenpl tl   o   a TS dop  teL a netlTf  ot  t   Sa  plSLe pe neft et Sd  p
pnL ete e f0d t 
    LSe  pf 0et0 e   d   
  0p   e00 e   d   
 0  0 0    0         0    00        0        0 0    0 0          0     0      0 0        0         00     0        -  0- 0 :   -0  -   :- - - : -  -00--  -:-- :--- :-  -:- --- ::0---  ---- :----::----:- ---:-::---  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 2552k  100 2552k    0     0  4550k      0 --:--:-- --:--:-- --:--:-- 4549k
100  475k  100  475k    0     0   139k      0  0:00:03  0:00:03 --:--:--  152k
100 1807k  100 1807k    0     0   398k      0  0:00:04  0:00:04 --:--:--  439k
100 32.9M    0 32.9M    0     0  5932k      0 --:--:--  0:00:05 --:--:-- 6718k
==> ERROR: A failure occurred in build().
    Aborting...
 -> error making: curl-impersonate-chrome-exit status 4
 -> Failed to install the following packages. Manual intervention is required:
curl-impersonate-chrome - exit status 4

**Comment by SuspiciousDuck** [comment]
What I noticed is that at the bottom, you can see that it is clearly just some malformed curl output. Specifically I noticed that they happened AFTER all of the unzip and tar errors. This means that the build instructions aren't waiting for the downloads. Aside from the download errors, that means I can't test to see if the errors are resolved.

**Comment by SuspiciousDuck** [comment]
That being said, now that I can actually test building curl-impersonate-chrome, I now see that no, there is still one more warning in BorringSSL, which errors the build. I think a possible solution is to just make another patch. You download the same archives, so why not just patch those archives to fix the build warnings?
This was the only warning I got.
```
[388/452] Building CXX object ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o
FAILED: ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o 
/usr/bin/g++ -DBORINGSSL_IMPLEMENTATION -I/home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/boringssl/third_party/googletest/include -I/home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/boringssl/ssl/test/../../include -Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wvla -Wshadow -ggdb -Wall -fvisibility=hidden -fno-common -Wno-free-nonheap-object -Wimplicit-fallthrough -Wmissing-declarations -std=c++11 -fno-exceptions -fno-rtti -O3 -DNDEBUG -fPIE -MD -MT ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o -MF ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o.d -o ssl/test/CMakeFiles/bssl_shim.dir/settings_writer.cc.o -c /home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/boringssl/ssl/test/settings_writer.cc
/home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/boringssl/ssl/test/settings_writer.cc: In member function ‘bool SettingsWriter::Commit()’:
/home/user/.cache/yay/curl-impersonate-chrome/src/curl-impersonate-0.6.0/build/boringssl/ssl/test/settings_writer.cc:78:61: error: ignoring attributes on template argument ‘int (*)(FILE*)’ [-Werror=ignored-attributes]
   78 |   using ScopedFILE = std::unique_ptr<FILE, decltype(&fclose)>;
      |                                                             ^
cc1plus: all warnings being treated as errors
```

**Comment by lwthiker** [comment]
>  tldr: -j12 ruins everything

Yes, parallel build doesn't work, it's a known issue which I haven't been able to solve yet.

> That being said, now that I can actually test building curl-impersonate-chrome, I now see that no, there is still one more warning in BorringSSL, which errors the build.

This warning was fixed in BoringSSL (https://boringssl.googlesource.com/boringssl/+/e4f60679caa293c047be69f57fc48b46c7452327). Try rebuilding: check out the branch #217, delete the build directory and run `make chrome-build` again.

**Comment by SuspiciousDuck** [comment]
Thanks a lot! Issue is resolved.
